src/machine.rs

```diff
@@ -120,10 +120,16 @@ impl fmt::Display for MiriMemoryKind {
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum AllocType {
+    Concrete(AllocId),
+    Casted,
+}
+
 /// Pointer provenance (tag).
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
 pub struct Tag {
-    pub alloc_id: AllocId,
+    pub alloc_id: AllocType,
```

Comment:
Actually, maybe this is not needed. `Pointer<Tag>` should be still fine with this PR, since `AllocId` has a niche. `Pointer<Option<Tag>>` will regress, but only temporarily -- once we remove `Untagged` from Stacked Borrows, we can hopefully use the niche in the SB tag to still get layout optimizations here. At least in theory; I am not sure how good Rust handles this in practice. (But we can cross that bridge when we get there. For this PR, as long as `Pointer<Tag>` stays in 24 bytes, I am happy.)